### ISSUE TITLE
Support eager bindings.

### ISSUE THREAD

**Issue by richburdon** [alecthomas/injector#44]
There doesn't seem to be a way to bind eager singletons. The only reference is in the Module's provides method, but this is unreferenced.

    def provides(interface, scope=None, eager=False):

**Comment by alecthomas** [comment]
Yes, I should remove that from the method signature. I don't think it will be added at this stage.

**Comment by richburdon** [comment]
Not added because it's not under active development or a time issue? Or not useful/practical?

BTW, are you still actively developing the project? Is there a better alternative?

Thanks.

**Comment by alecthomas** [comment]
Mostly the latter; not useful enough to warrant the effort IMO.

Yes, Injector is still being developed, it's just mostly stable so there's not much activity.

You can have a look at [pinject](https://github.com/google/pinject), but it hasn't been touched in two years and IMO is not as nice as Injector anyway ;)

**Comment by jstasiak** [comment]
I can confirm that the lack of activity is due to the code just working.

**Comment by richburdon** [comment]
Fair enough :).

I find myself in a minority of Python developers who can't live without DI,
so thanks for this package.

I do however, think eager singletons are important. As would like to see
Sequence and Map keys support instantiation by class (rather than by
instance). I'll try to send a pull request for the workarounds that I have
created.

**Comment by juharris** [comment]
On a similar note, I'd like to be able instantiate **all** singletons when my server starts, something analogous to Guice's [Stage.PRODUCTION](https://google.github.io/guice/api-docs/latest/javadoc/com/google/inject/Stage.html#PRODUCTION).  Any particular reasons why it's not supported? Would you mind if I added it? I'd probably do it as a `stage` parameter to `Injector`.

**Comment by jstasiak** [comment]
I completely forgot this issue was still open :) I wouldn't be against it in principle, I'm just wondering what the actual implementation would look like.

**Comment by jstasiak** [comment]
Yeah I'm not saying it's not useful, it's the implementation I wondered about, the general case seems tricky. For example:

```python

@singleton
SomeClass:
    pass


injector = Injector()
eagerly_instantiate_singletons(injector)
```

With `auto_bind` enabled (the default) Injector doesn't know anything about `SomeClass` until it's instantiated so the eager instantiation mechanism would need to actively look for classes (All loaded classes? Classes in a particular module?).

Depending on the application the easiest and the most helpful could be to do

```python
for t in [list, of, some, known, singletons, that, can, cause, trouble]:
    injector.get(t)
```

at the start of your program.

BTW not sure if memory saving would work in the scenario you describe, see https://instagram-engineering.com/dismissing-python-garbage-collection-at-instagram-4dca40b29172

**Comment by juharris** [comment]
So for Guice it only knows about the singletons mentioned in the module or through dependencies. It would be a great start to instantiate those.

**Comment by Sytten** [comment]
I found myself wanting this feature quite badly. I use Injector to inject my service layer in my flask (flask-rebar to be precise) controllers. Some services need to fetch external dependencies, so the first call to the API takes a big hit in performances. I will try the trick of forcing the injector to get the type, but an `eager` parameter in the `@singleton` decorator would be much cleaner.

**Comment by levsa** [comment]
I would like to request a feature like this as well. I would like to be able to annotate singletons so that they are automatically instantiated upon creation of the injector.

The reason is that I have a module that is to be included by many injectors that have different set of modules, and it would be nice to not need to do `injector.get(MySingleton)` from all the places where injectors are created.

**Comment by delucca** [comment]
in the meantime, I would like to share my solution:

```python
from typing import Callable

from injector import Binder


def eager_initialize(eager_modules: list):
    def decorate_function(configure: Callable):
        def wrapper(self, binder: Binder, *args, **kwargs):
            result = configure(self, binder, *args, **kwargs)

            for module in eager_modules:
                binder.injector.get(module)

            return result

        return wrapper

    return decorate_function
```

You can use it as a decorator in your configure `Module` method, like this:

```python
class RequestedManufacturerNameModule(Module):
    @eager_initialize([RequestedManufacturerNameService])
    def configure(self, binder: Binder) -> None:
        binder.bind(JobHandler, to=JobHandlerFactory.from_app, scope=singleton)
        binder.bind(DataframeFactory, scope=singleton)
        binder.bind(
            DataframeTransformer,
            to=RequestedManufacturerNameDataframeTransformer,
            scope=singleton,
        )
        binder.bind(RequestedManufacturerNameService, scope=singleton)
        binder.bind(
            RequestedManufacturerNameFeatureExtractor,
            to=RequestedManufacturerNameFeatureExtractor.from_app,
            scope=singleton,
        )
        binder.bind(
            RequestedManufacturerNameClassifier,
            to=RequestedManufacturerNameClassifier.from_app,
            scope=singleton,
        )
``` 

You can pass a list of eager classes to it and it should work. But, in any case, it would be way better doing so in the `binder` directly.

**Comment by sheinbergon** [comment]
@jstasiak would you be willing to accpet @delucca implementation as a PR (including documentation update)?

**Comment by sheinbergon** [comment]
@jstasiak?

**Comment by davidparsson** [comment]
I'm not the primary maintainer of this library, but I would prefer this to be implemented as a Scope (if possible) instead of a custom decorator. It's more in line with how the rest of the library works, and it is also more simliar to how Guice has solved it. But there might be problems with that which I'm not forseeing.

**Comment by sheinbergon** [comment]
@davidparsson is anyone still maintaing this library?

**Comment by jstasiak** [comment]
Yeah I still maintain it, I'll probably get around to a round of code review etc. soon.

In this case my silence is because of a combination of:

* I'm truly not sure what the best design for this
* I never needed it
* It seems that if one needs that it's fairly easy to implement it in the client code exactly the way one needs it to work so it's not a high priority, subjectively speaking

**Comment by sheinbergon** [comment]
@jstasiak Thank you for responding.
The use case is quite simple - initailize on startup rather than wait for API access, it helps flush wiring / initialization bugs earlier, which is crucial for keeping the deployment process safe and sane.
 
 If I'll open a PR for an @eager decorator (not the current design, but a per @singleton/@provider function), would you be willing to accept that?

**Comment by davidparsson** [comment]
@sheinbergon, with the proper interface I think this can be a useful addition. I have actually been wanting this myself, although only once.

It is pretty straightforward to implement eager singletons in a way that works, but sometimes, and especially in large code bases with many different Injectors, I think it makes sense to push the definition of what needs to be eager close to the implementation of the singleton rather than in an Injector module. An `@eager` decorator would do just that.

(And, sorry for my slow replies. I have a hard time keeping up with life at the moment.)

**Comment by davidparsson** [comment]
@sheinbergon, please go ahead and give it a try! [This](https://github.com/python-injector/injector/issues/44#issuecomment-436665244) might still be a problem though.

**Comment by davidparsson** [comment]
All right, that’s reasonable. In that case I think I’d prefer `@provider(eager=True)` over a separate decorator. Thoughts?

**Comment by sheinbergon** [comment]
Yes, I think we should parameterize `@singleton` , as `@multiprovider`  can also be made eager
